rp2040-hal/src/pio.rs

```diff
@@ -165,57 +185,175 @@ impl<P: Instance> PIO<P> {
+///
+/// Objects of this type can be reused for multiple state machines of the same PIO block to save
+/// memory if multiple state machines are supposed to perform the same function (for example, if
+/// one PIO block is used to implement multiple I2C busses).
+///
+/// `PIO::uninstall(program)` can be used to free the space occupied by the program once it is no
+/// longer used.
+///
+/// # Examples
+///
+/// ```
+/// let (mut pio, sm0, _, _, _) = pac.PIO0.split(&mut pac.RESETS);
+/// // Install a program in instruction memory.
+/// let installed = pio.install(&program).unwrap();
+/// // Configure a state machine to use the program.
+/// let sm = rp2040_hal::pio::PIOBuilder::from_program(installed).build(sm0);
+/// // Uninitialize the state machine again, freeing the program.
+/// let (sm, installed) = sm.uninit();
+/// // Uninstall the program to free instruction memory.
+/// pio.uninstall(installed);
+/// ```
+///
+/// # Safety
+///
+/// Objects of this type can outlive their `PIO` object. If the PIO block is reinitialized, the API
+/// does not prevent the user from calling `uninstall()` when the PIO block does not actually hold
+/// the program anymore. The user must therefore make sure that `uninstall()` is only called on the
+/// PIO object which was used to install the program.
+///
+/// ```
+/// let (mut pio, sm0, sm1, sm2, sm3) = pac.PIO0.split(&mut pac.RESETS);
+/// // Install a program in instruction memory.
+/// let installed = pio.install(&program).unwrap();
+/// // Reinitialize PIO.
+/// let pio0 = pio.free(sm0, sm1, sm2, sm3);
+/// let (mut pio, _, _, _, _) = pio0.split(&mut pac.RESETS);
+/// // Do not do the following, the program is not in instruction memory anymore!
+/// pio.uninstall(installed);
+/// ```
+#[derive(Debug)]
+pub struct InstalledProgram<P> {
+    offset: u8,
+    length: u8,
+    side_set: SideSet,
+    wrap: Wrap,
+    _phantom: core::marker::PhantomData<P>,
+}
+
+impl<P: PIOExt> InstalledProgram<P> {
+    /// Clones this program handle so that it can be executed by two state machines at the same
+    /// time.
+    ///
+    /// # Safety
+    ///
+    /// This function is marked as unsafe because, once this function has been called, the
+    /// resulting handle can be used to call `PIO::uninstall()` while the program is still running.
+    ///
+    /// The user has to make sure to call `PIO::uninstall()` only once and only after all state
+    /// machines using the program have been uninitialized.
+    pub unsafe fn share(&self) -> InstalledProgram<P> {
+        InstalledProgram {
+            offset: self.offset,
+            length: self.length,
+            side_set: self.side_set,
+            wrap: self.wrap,
+            _phantom: core::marker::PhantomData,
         }
     }
 }
 
-/// PIO State Machine.
+/// PIO State Machine (uninitialized, without a program).
 #[derive(Debug)]
-pub struct StateMachine<P: Instance> {
+pub struct UninitStateMachine<P: PIOExt> {
```

Comment:
I'd like to see this be `StateMachine<P, Uninit>`, but when I tried to draft an example of that I realized we probably need GATs for that because `InstalledProgram` is generic over `P`. Another option we could have today is
```rust
StateMachine<Uninit<PIO0>>;
StateMachine<StoppedPIO0>>;
StateMachine<Starged<PIO0>>;
```
but I don't like that either.

Reply:
I tried the latter, but did not like that it made it more difficult to implement functions that can operate on `Stopped` and `Running`, but not on `Uninit`. There are quite some such functions that just do not make sense on an UninitStateMachine.

I do not see this as a substantial usability concern, but I can look further into it if you insist.

Reply:
I see. I guess one could still have something like
```rust
trait Initialized {...}
impl Initialized for Stopped {...}
impl Initialized for Running {...}
impl<State: Initialized> StateMachine<State> {...}
```
if one wanted to implement something only for `Stopped` and `Running`. But I don't insist on changing this, I'm just trying to brainstorm some ideas on how to make the API more self-consistent.